### DISCUSSION TITLE
Are there any plans to replace patching by delegating?

### DISCUSSION THREAD

**Discussion by zby** [567-labs/instructor#301] [Ideas]
For some time we have the openai client object - so instead of patching the library we could just store the openai client object, use it to make calls and then process the response.

The example would become:
```
instructor = Instructor()
user = instructor.process(
    model="gpt-3.5-turbo",
    response_model=UserDetail,
    messages=[
        {"role": "user", "content": "Extract Jason is 25 years old"},
    ]
)
```
When creating the Instructor one could also pass a client - this would be useful for tests.

**Comment by jxnl** [comment]
honestly know, i'm pretty hesitant to introduce a new sdk to wrap an existing sdk. its another layer of abstraction i didn't earn

**Reply by zby** [reply]
So maybe inheritance? It is like patching but a bit more civilized.
The disadvantage would be losing the "don't mess with me" warning.

But I expect that once you add other LLMs you'll have to make a new api anyway.

**Reply by jxnl** [reply]
[not that interested right now](https://jxnl.github.io/instructor/concepts/philosophy/) im sure 95% of all llm api calls are either openai or mirror the openai sdk. 
if people want to use code the integrates with every single api they can use 

https://python.langchain.com/docs/modules/model_io/output_parsers/pydantic 
https://www.askmarvin.ai/
https://gpt-index.readthedocs.io/en/latest/examples/output_parsing/openai_pydantic_program.html

**Reply by jxnl** [reply]
I think inheritance makes sense from the perspective of getting better IDE support for the new word arguments. 

but having `Instructor()` 
means i now have to support whatever wide range of key word arguments from any client, rather than letting OpenAI() handle it. azure, openai. mistral. together. etc

you could image class methods 

```
instructor.from_openai
instructor.from_mistral
instructor.from_ollama 
```

but its a lot of work to support 1% of the user base, esp when each model has different prompts, the library starts being responsible for writing prompts rather than just supporting the function calling and tool calling apis
